### ISSUE TITLE
BSON decoder error

### ISSUE THREAD

**Issue by noam87** [elixir-mongo/mongodb#59]
Hi, on one of my collections I get the following error:

```
 ** (FunctionClauseError) no function clause matching in BSON.Decoder.type/2
            (mongodb) lib/bson/decoder.ex:10: BSON.Decoder.type(1, <<1, 0, 0, 0, 0, 0, 240, 127, 1, 49, 0, 1, 0, 0, 0, 0, 0, 240, 127>>)
```

```
 lib/bson/decoder.ex:10: BSON.Decoder.type(1, <<0, 0, 0, 0, 0, 0, 240, 127, 1, 112, 114, 111, 106, 101, 99, 116, 101, 100, 69, 84, 99, 70, 114, 111, 109, 104, 105, 115, 116, 111, 114, 121, 0, 170, 241, 210, 77, 98, 16, 201, 63, 2, 67, 73, 77, 73, 83, 83, 116, 97, ...>>)
```

I am pulling from master since it seems a similar issue was resolved recently, but it still breaks. I'm connecting to an existing database so I need to be able to access the values. Is there a workaround for this type of issue before it is fixed?

I think it is caused by a 2 element array that has `nan`s with type`double`.

Thanks.

**Comment by noam87** [comment]
I think it is caused by a 2 element array that has `nan`s with type`double`.

**Comment by noam87** [comment]
Hm, it's happening in a different collection too, which does not have any `nan`s:

```
 lib/bson/decoder.ex:10: BSON.Decoder.type(1, <<0, 0, 0, 0, 0, 0, 240, 127, 1, 112, 114, 111, 106, 101, 99, 116, 101, 100, 69, 84, 99, 70, 114, 111, 109, 104, 105, 115, 116, 111, 114, 121, 0, 170, 241, 210, 77, 98, 16, 201, 63, 2, 67, 73, 77, 73, 83, 83, 116, 97, ...>>)
```

**Comment by ericmj** [comment]
It looks like the bytes are reversed compared to what we expect. You have `<<0, 0, 0, 0, 0, 0, 240, 127>>`, we expect `<<127, 240, 0, 0, 0, 0, 0, 0, 0>>`.

The byte layout we expect is IEEE 754 so my first guess would be that there is an issue with the library that inserted the values. Do you know what library that is?

**Comment by noam87** [comment]
Thanks for the help.

The db value were inserted via node.js app I've inherited (and rewriting in Elixir). 
- It's using the library `mongoose`.
- The values in question come from the clients and converted from string to numbers using node's native `Number(value)`.

**Comment by noam87** [comment]
Also, other stored values look correct and can be accessed successfully.

**Comment by ericmj** [comment]
Pinging @TreyE who added the support for inf and nan.

**Comment by TreyE** [comment]
Just saw this, I'm currently examining the necessity of doing byte-order examination on IEE 754 values.  It might be I flipped the bits, or that I need to check the endianness before attempting to decode.  It appears my bit-matching pattern may be missing an endianness specification.

According to the BSON specification, each type must be formatted using little-endian storage.  I wrote the matchers in a big-endian manner it seems.  Let me reverse them and see what happens.
